Fix YAML's default_value_if handling

bkaestner opened the pull request:
While working on another issue I was completely puzzled by the `yaml_opt_str!`, as it immediately starts with `$v.is_null()` and then uses `as_str()` on a `YAML::Null` value. While trying to get to the root of the issue, I've noticed that this bug will only manifest itself in `default_value_if` when compiled with `yaml` support.

I've added some integration tests for the YAML part and also noticed that `null` was misspelled in the fixture, which kept the bug hidden. If you don't apply `e20cc1c` from the set of commits, you'll notice that *all* `yaml` tests break down due to `[flag, null, some]`.

**Please note that this is my first non-documentation Rust pull request ever and I'm happy about any feedback and criticism.**

I wasn't sure about the commit message style, as the recent commits didn't use the style mentioned in the contributing guide. I've tried to emulate the style from the recent commits. Also wasn't sure whether you want to have a single squashed commit, or three atomic changes/additions, but I'll happily change the messages/squash the commits if necessary.

pksunkara commented:
> I wasn't sure about the commit message style, as the recent commits didn't use the style mentioned in the contributing guide. I've tried to emulate the style from the recent commits.

Yeah, we haven't been following conventional commit style since we are working on v3. We will probably decide on this after v3 is released.

> Also wasn't sure whether you want to have a single squashed commit, or three atomic changes/additions, but I'll happily change the messages/squash the commits if necessary.

Your commits are pretty good. They each deal with different stuff and your messages are self-explanatory. I think it's okay to have them like. I find that it's always a balancing issue.

But if you have some tests failing, you will probably need to fix that tests by amending the relevant commit.

bkaestner commented:
> Your commits are pretty good. 

Thank you!

> But if you have some tests failing, you will probably need to fix that tests by amending the relevant commit.

Yeah, just noticed that and completely messed that up by testing two things at once, my bad. I missed the `ArgRequiredElseHelp` in the YAML while I simultaneously hit the `Null` issue in the fixtures and added another WRONG test for that. I removed the bad test (and this time also checked the correct commit locally).

Sorry about the wrong test :/

pksunkara commented:
bors r+

bkaestner commented:
@pksunkara I just noticed that [`v2-master` also contains this bug](https://github.com/clap-rs/clap/blob/v2-master/src/args/macros.rs#L68). I can backport the fix as well as the test cases into v2 if necessary.

pksunkara commented:
We only do security fixes for v2
